### PR TITLE
Mysky autologin stored settings and logout

### PR DESCRIPTION
**Pull request by m-cat**
# PULL REQUEST

## Overview

- Implements logout
- Implements auto re-login
- Implements getting user settings such as preferred portal and email
- Implements most of the preferred portal flow
- Moves portal login from MySky UI to Main MySky

## Portal refresh flow

```ts
/**
*  Load MySky redirect flow:
*
*  1. SDK opens MySky on the same portal as the skapp.
*  2. If the preferred portal is found in localstorage, MySky connects to it
*     and we go to step 5.
*  3. Else, MySky connects to siasky.net.
*  4. MySky tries to get the saved portal preference.
*     1. If the portal is set, MySky switches to using the preferred portal.
*     2. If it is not set or we don't have the seed, MySky switches to using
*        the current portal as opposed to siasky.net.
*  5. After MySky finishes loading, SDK queries `mySky.getPortalPreference`.
*  6. If the preferred portal is set and different than the current portal,
*     SDK triggers refresh.
*  7. We go back to step 1 and repeat, but since we're on the right portal
*     now we won't refresh in step 6.
*
* Login redirect flow:
*
* 1. SDK logs in through the UI.
* 2. MySky UI switches to siasky.net and tries to get the saved portal
*    preference.
*    1. If the portal is set, MySky switches to using the preferred portal.
*    2. If it is not set or we don't have the seed, MySky switches to using
*       the current portal as opposed to siasky.net.
* 3. SDK queries `mySky.getPortalPreference`.
* 4. If the preferred portal is set and different than the current portal,
*    SDK triggers refresh.
* 5. We go to "Load MySky" step 1 and go through that flow, but we don't
*    refresh in step 6.
*/
```

## TODO

- [ ] Figure out how to get stored data to persist across portals during redirect. ( see below )
- [x] Still have to signal to MySky UI that we are done logging in.
- [ ] *Add dedicated flow(s) for changing the email after it's set. (in a future PR)*

These TODOs are also documented in comments in the code.

### Persisting data across portals

- [ ] Persist settings from one portal to another when redirecting
  - [x] Check if email is valid first by trying to register/login
  - [x] Save everything in user settings
  - [x] Redirect
  - [x] Load email from user settings on MySky load and login again
  - [x] Save seed/email/preferred portal in local storage
  - [ ] ******* How to send seed to MySky on new portal? *******

### Testing

Of course this needs to be tested as well.

- [ ] Test register + login flows
- [ ] Make a skapp that sets preferred portal
- [ ] Test preferred portal redirect

## SDK PR

See https://github.com/SkynetLabs/skynet-js/pull/361, which implements the page refresh in the SDK. These must be tested in tandem.

### CODE REVIEW

File: src/mysky.ts [hunk 1]
```diff
@@ -345,31 +685,67 @@ export class MySky {
...
+          // signed up with it -- check local storage. We don't need to do this if
+          // the email was already found.
+          // TODO: Add dedicated flow(s) for changing the email after it's set.
+          let isEmailProvided = false;
+          if (!storedEmail) {
+            storedEmail = checkStoredEmail();
+            isEmailProvided = storedEmail !== null;
+          }
+
+          if (storedEmail) {
```
[the remark this answers is not in the document]
Comment (peterjan) [reply src/mysky.ts:718]: this saves you a bunch of lines and a boolean, in any case I would rename `isEmailProvided` to something that explains why it will be used, or update the user settings inside of the `if (!storedEmail) {`, both would read easier

File: src/mysky.ts [hunk 2]
```diff
@@ -1,26 +1,42 @@
...
   signEntry,
   SkynetClient,
   PUBLIC_KEY_LENGTH,
   PRIVATE_KEY_LENGTH,
+  RequestConfig,
+  ExecuteRequestError,
+  JsonData,
+  deriveEncryptedFileKeyEntropy,
+  EncryptedJSONResponse,
+  MAX_REVISION,
```
[the remark this answers is not in the document]
Comment (m-cat) [reply src/mysky.ts:21]: You need to link to the dependency PR mentioned in the description.

File: src/mysky.ts [hunk 3]
```diff
@@ -48,74 +72,106 @@ export class PermissionsProvider {
...
+    const referrerDomain = await actualPortalClient.extractDomain(document.referrer);
 
     // Create MySky object.
-
     log("Calling new MySky()");
-    const mySky = new MySky(client, referrerDomain, seed);
+    const mySky = new MySky(initialClient, mySkyDomain, referrerDomain, permissionsProvider, preferredPortal);
+
+    // Login to portal.
+    {
```
[the remark this answers is not in the document]
Comment (m-cat) [reply src/mysky.ts:150]: I thought it made it more clear that the comment applies to the whole scope.

File: src/mysky.ts [hunk 4]
```diff
@@ -325,18 +417,266 @@ export class MySky {
...
-   * Set up a listener for the storage event. If the seed is set in the UI, it
-   * should trigger a load of the permissions provider.
+   * Sets up the handshake connection with the parent.
+   */
+  protected connectToParent(): void {
+    // Set child methods.
+
+    const methods = {
+      checkLogin: this.checkLogin.bind(this),
+      getEncryptedFileSeed: this.getEncryptedPathSeed.bind(this),
```
[the remark this answers is not in the document]
Comment (m-cat) [reply src/mysky.ts:427]: Added a comment.

File: src/mysky.ts [hunk 5]
```diff
@@ -325,18 +417,266 @@ export class MySky {
...
+    validateObject("json", json, "parameter");
+
+    const opts = { hashedDataKeyHex: true };
+
+    // Call MySky which checks for read permissions on the path.
+    const [publicKey, pathSeed] = await Promise.all([this.userID(), this.getEncryptedPathSeed(path, false)]);
+    const dataKey = deriveEncryptedFileTweak(pathSeed);
+
+    // Immediately fail if the mutex is not available.
+    return await this.client.db.revisionNumberCache.withCachedEntryLock(
```
[the remark this answers is not in the document]
Comment (m-cat) [reply src/mysky.ts:550]: This PR assumes the SkyDB rework has been merged in. The dependency PR is built on top of the latest changes in skynet-js.

File: src/mysky.ts [hunk 6]
```diff
@@ -345,31 +685,67 @@ export class MySky {
...
-        localStorage.removeItem(EMAIL_STORAGE_KEY);
-
-        // TODO: Set up auto-login.
+        // NOTE: Don't use the stored preferred portal here because we are just
+        // logging into a new account and need to get the user settings for the
+        // first time. Always use siasky.net.
+        this.client = getLoginClient(seed, null);
+
+        // Login to portal.
+        {
```
[the remark this answers is not in the document]
Comment (m-cat) [reply src/mysky.ts:700]: Yes, it's a block scope.

File: src/mysky.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (m-cat) [reply src/mysky.ts:718]: Renamed the variable and added an explanatory comment.

File: src/mysky.ts [hunk 7]
```diff
@@ -325,18 +417,266 @@ export class MySky {
...
+   * ahead of time, and an error would not be actionable.
+   *
+   * NOTE: We restore the original `executeRequest` on logout. We do not try to
+   * modify `executeRequest` if it is already modified and throw an error
+   * instead.
+   *
+   * @param seed - The user seed.
+   * @param email - The user email.
+   */
+  protected setupAutoRelogin(seed: Uint8Array, email: string): void {
```
[the remark this answers is not in the document]
Comment (m-cat) [reply src/mysky.ts:624]: I tried moving this logic into `executeRequest` as you suggested and it didn't seem like an improvement. If anything it just spread the logic out over two projects and coupled `skynet-js` even more closely with MySky. The way it is, the logic is confined to one place, and though it is slightly confusing I've seen much more confusing hook/callback type logic in Javascript...

File: src/mysky.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (m-cat) [reply src/mysky.ts:704]: Was this confusing to you? I googled to see if it was standard and some people on StackOverflow were confused by it. Not sure how seriously to take that, though, because people on the internet have different experience levels. To my eyes this looks pretty clear. I wanted to avoid extracting this into a new function because having a lot of nested functions can also be hard to follow.

File: src/mysky.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (m-cat) [reply src/mysky.ts:624]: I was debugging this code and I realized that your suggestion would actually simplify it (as well as fix my bug -- think I was missing a `.bind(this)` somewhere but that's not needed if the logic is already in `client.executeRequest`.) I pushed the change in the other PR. I called it `loginFn` though to hopefully not couple this too closely with "MySky auto-relogin".

**Comment by ro-tex** [comment]
This is not exactly a code problem but we need to stick to *some* naming convention. Right now we have a bunch of kebab-case and a bunch of snake_case filenames, sometimes in the same directory.

**Comment by m-cat** [comment]
@ro-tex Braised this for the file-naming issue: https://github.com/SkynetLabs/skynet-mysky/issues/241

File: src/mysky.ts [hunk 8]
```diff
@@ -507,24 +839,104 @@ export function checkStoredSeed(): Uint8Array | null {
...
+export async function getCurrentAndReferrerDomains(): Promise<{
+  currentDomain: string;
+  referrerDomain: string | null;
+}> {
+  // Get the MySky domain (i.e. `skynet-mysky.hns or sandbridge.hns`). Use
+  // hard-coded values since we don't expect official MySky to be hosted
+  // anywhere else for now.
+  let currentDomain;
+  if (ALPHA_ENABLED && DEV_ENABLED) {
+    throw new Error("Alpha and dev modes cannot both be enbaled");
```
Comment (ro-tex) [inline src/mysky.ts:891]:
```suggestion
    throw new Error("Alpha and dev modes cannot both be enabled");
```

File: src/mysky.ts [hunk 9]
```diff
@@ -199,54 +236,108 @@ export class MySky {
...
   /**
    * Logs out of MySky.
    */
   async logout(): Promise<void> {
-    // Clear the stored seed.
+    const errors = [];
 
-    clearStoredSeed();
+    // Check if user is logged in.
+    const seed = checkStoredSeed();
```
Comment (ro-tex) [inline src/mysky.ts:273]: nit: We are doing this operation multiple times in this file. Can we avoid that by caching the seed somewhere?

File: src/mysky.ts [hunk 10]
```diff
@@ -374,18 +463,241 @@ export class MySky {
...
+    // NOTE: Don't use the stored preferred portal here because we are just
+    // logging into a new account and need to get the user settings for the
+    // first time. Always use siasky.net.
+    this.client = getLoginClient(seed, null);
+
+    // Login to portal.
+    {
+      // Get the preferred portal and switch to it.
+      const { portal: preferredPortal, email } = await getUserSettings(this.client, seed, this.mySkyDomain);
+      let storedEmail = email;
```
Comment (ro-tex) [inline src/mysky.ts:563]:
```suggestion
      const { portal: preferredPortal, email: storedEmail } = await getUserSettings(this.client, seed, this.mySkyDomain);
```

File: src/mysky.ts [hunk 11]
```diff
@@ -394,26 +706,23 @@ export class MySky {
...
+      localStorage.removeItem(PORTAL_LOGIN_COMPLETE_SENTINEL_KEY);
+
+      try {
+        // Parse the seed.
+        const seed = new Uint8Array(JSON.parse(newValue));
+
+        await this.loginFromUi(seed);
+
+        // Signal to MySky UI that we are done.
+        localStorage.setItem(PORTAL_LOGIN_COMPLETE_SENTINEL_KEY, "1");
```
Comment (ro-tex) [inline src/mysky.ts:720]: nit: This `1` is the only acceptable value for this key. That's confusing. It would be better to put it in a const, so we know that there is only one valid sentinel value.

File: src/provider.ts [hunk 12]
```diff
@@ -48,7 +49,7 @@ export async function getPermissionsProviderUrl(_seed: Uint8Array): Promise<stri
  * @returns - The handshake connection with the provider.
  */
 export async function launchPermissionsProvider(seed: Uint8Array): Promise<PermissionsProvider> {
-  console.log("Entered launchPermissionsProvider");
+  log("Entered launchPermissionsProvider");
```
[the remark this answers is not in the document]
Comment (m-cat) [reply src/provider.ts:52]: Wrote a comment in the code and braised an issue:
https://github.com/SkynetLabs/skynet-mysky/issues/242

File: src/mysky.ts (hunk 9 above)
Comment (m-cat) [reply src/mysky.ts:273]: Yes, that is indeed a good point. I wrote up an issue: https://github.com/SkynetLabs/skynet-mysky/issues/243

File: src/skydb_internal.ts [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+ * @returns - The incremented revision number.
+ * @throws - Will throw if the incremented revision number is greater than the maximum revision.
+ */
+function incrementRevision(revision: bigint): bigint {
+  revision = revision + BigInt(1);
+
+  // Throw if the revision is already the maximum value.
+  if (revision > MAX_REVISION) {
+    throw new Error("Current entry already has maximum allowed revision, could not update the entry");
+  }
```
[the remark this answers is not in the document]
Comment (m-cat) [reply src/skydb_internal.ts:141]: Sorry, I don't understand why this is an issue. 😅 Can you explain further?

File: src/skydb_internal.ts [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+ * @returns - The hex-encoded encrypted path seed.
+ */
+async function getEncryptedPathSeedInternal(seed: Uint8Array, path: string, isDirectory: boolean): Promise<string> {
+  log("Entered getEncryptedPathSeedInternal");
+
+  // Compute the root path seed.
+  const rootPathSeedBytes = deriveRootPathSeed(seed);
+
+  // Compute the child path seed.
+  return deriveEncryptedPathSeedForRoot(rootPathSeedBytes, path, isDirectory);
```
[the remark this answers is not in the document]
Comment (m-cat) [reply src/skydb_internal.ts:124]: I see your point, as `deriveEncryptedPathSeedForRoot` here can only be used with the root seed. I would like to keep it this way though so that `deriveEncryptedPathSeedForRoot` is similar to `deriveEncryptedPathSeed` in skynet-js with minimal changes. `skynet-js` contains the reference implementation and I think it's more clear if we try to stick to it with as little changes as possible.

File: src/skydb_internal.ts [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+  validateString("path", path, "parameter");
+  validateObject("json", json, "parameter");
+
+  const { publicKey } = genKeyPairFromSeed(seed);
+  const pathSeed = await getEncryptedPathSeedInternal(seed, path, false);
+  const dataKey = deriveEncryptedFileTweak(pathSeed);
+  const opts = { hashedDataKeyHex: true };
+
+  // Immediately fail if the mutex is not available.
+  return await client.db.revisionNumberCache.withCachedEntryLock(publicKey, dataKey, async (cachedRevisionEntry) => {
```
[the remark this answers is not in the document]
Comment (m-cat) [reply src/skydb_internal.ts:83]: I don't think we should export `CachedRevisionNumber` from `skynet-js` as it's an internal implementation detail, and MySky needing to access it is an unusual case. I added a type annotation though.

File: src/skydb_internal.ts [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+
+import { deriveRootPathSeed, genKeyPairFromSeed } from "./crypto";
+import { deriveEncryptedPathSeedForRoot } from "./encrypted_files";
+import { log, validateObject, validateString } from "./util";
+
+/**
+ * Gets Encrypted JSON at the given path through MySky.
+ *
+ * @param client - The Skynet client.
+ * @param seed - The user seed.
```
[the remark this answers is not in the document]
Comment (m-cat) [reply src/skydb_internal.ts:24]: That's a good question, indeed. It will be the MySky root seed and not an account-specific seed. I will update the relevant docstrings to make this distinction more clear for future readers of the code.

File: package.json [hunk 17]
```diff
@@ -45,13 +45,14 @@
...
     "confusables": "^1.1.0",
+    "core-js": "^3.20.2",
     "crypto-browserify": "^3.12.0",
     "idb-keyval": "^6.1.0",
     "mustache": "^4.2.0",
     "post-me": "^0.4.5",
     "punycode": "^2.1.1",
     "randombytes": "^2.1.0",
-    "skynet-js": "^4.0.20-beta",
+    "skynet-js": "^4.0.22-beta",
```
[the remark this answers is not in the document]
Comment (m-cat) [reply package.json:55]: This doesn't matter for now because I've linked to a specific branch in `skynet-js` locally. Changing this line wouldn't affect anything, I just have to keep the `skynet-js` branch up-to-date with recent changes. I will change this line once the `skynet-js` branch has been merged and a version change published.

File: src/skydb_internal.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (ro-tex) [reply src/skydb_internal.ts:145]: Oh, my bad. I am automatically thinking in pointers. :)

File: src/mysky.ts (hunk 10 above)
Comment (m-cat) [reply src/mysky.ts:564]: Making both of them mutable results in a lint error, since one of them is never mutated.

File: scripts/ui.ts [hunk 18]
```diff
@@ -494,6 +464,38 @@ async function setupAndRunDisplay<T>(displayUrl: string, methodName: string, ...
     });
 }
 
+/**
+ * Resolves when portal login on Main MySky completes.
+ *
+ * @returns - An empty promise.
+ */
+async function resolveOnMySkyPortalLogin(): Promise<void> {
```
[the remark this answers is not in the document]
Comment (m-cat) [reply scripts/ui.ts:472]: I will happily add more clarifying comments, thank you for the suggestion!

Regarding question 2: local storage is sandboxed per domain, so `badapp.hns.siasky.net` can't write to MySky's local storage. And MySky should only ever run its own code. If this invariant is ever broken, then yes: we're dead.

[the remark this answers is not in the document]
Comment (ro-tex) [reply scripts/ui.ts:479]: Thanks for explaining! One day I'll read all those fascinating things about browsers - storage types, workers, all of it. Until then I'll keep asking silly questions. ;)

File: src/mysky.ts (hunk 10 above)
Comment (ro-tex) [reply src/mysky.ts:564]: Ha ha, ok, it's settled then. Let's keep it as it is.

File: scripts/ui.ts [hunk 19]
```diff
@@ -494,6 +465,61 @@ async function setupAndRunDisplay<T>(displayUrl: string, methodName: string, ...
...
+  log("Entered resolveOnMySkyPortalLogin");
+
+  const abortController = new AbortController();
+
+  // Set up a promise that succeeds on successful login in main MySky, and fails
+  // when the login attempt returns an error.
+  const promise1 = new Promise<void>((resolve, reject) => {
+    const handleEvent = async ({ key, newValue }: StorageEvent) => {
+      if (key !== PORTAL_LOGIN_COMPLETE_SENTINEL_KEY) {
+        return;
```
Comment (peterjan) [inline scripts/ui.ts:489]: so this makes it so `promise1` never resolves, are we good with that?

File: scripts/ui.ts [hunk 20]
```diff
@@ -494,6 +465,61 @@ async function setupAndRunDisplay<T>(displayUrl: string, methodName: string, ...
...
+      }
+      if (!newValue) {
+        // Key was removed.
+        return;
```
Comment (peterjan) [inline scripts/ui.ts:493]: same here

File: src/mysky.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (peterjan) [reply src/mysky.ts:149]: hm ok, kind of implies it should/could be a separate function but equally fine leaving this in I guess 🤷‍♂️

File: src/skydb_internal.ts [hunk 21]
```diff
@@ -0,0 +1,194 @@
...
+  validateString("path", path, "parameter");
+  validateObject("json", json, "parameter");
+
+  const { publicKey } = genKeyPairFromSeed(seed);
+  const pathSeed = await getEncryptedPathSeedInternal(seed, path, false);
+  const dataKey = deriveEncryptedFileTweak(pathSeed);
+  const opts = { hashedDataKeyHex: true };
+
+  // Immediately fail if the mutex is not available.
+  return await client.db.revisionNumberCache.withCachedEntryLock(
```
Comment (peterjan) [inline src/skydb_internal.ts:83]: do we care about these linting errors here? `Property 'revisionNumberCache' does not exist on type` looks li ke the database isn't properly typed here

**Review by peterjan — APPROVED** [review]
I'm not fully up to speed with the feature, but from a pure code standpoint this PR looks solid and I don't have any blocking comments, LGTM 👍

File: html/seed-display.html [hunk 22]
```diff
@@ -111,7 +111,7 @@ <h2 class="mt-6 text-center text-3xl font-extrabold text-gray-900">
       <!-- Signup form -->
       <div class="mt-8 sm:mx-auto sm:w-full sm:max-w-md">
         <div class="bg-white py-8 px-4 shadow sm:rounded-lg sm:px-10">
-          <form class="space-y-6" action="#" onsubmit="window.signUp()">
+          <form class="space-y-6" action="#" onsubmit="window.signUp(event)">
```
[the remark this answers is not in the document]
Comment (m-cat) [reply html/seed-display.html:114]: That sounds familiar to me, though I unfortunately don't understand what that does either. However,

1. I know that everything works the way it is now,
2. I would have to do all the deploys and testing again after making this change, and this was pretty annoying to test.

To me it makes sense to move this to a followup. Maybe at some point I can do a blitz through Issues with Karol or Farzad.

[the remark this answers is not in the document]
Comment (m-cat) [reply html/seed-display.html:114]: https://github.com/SkynetLabs/skynet-mysky/issues/252

File: scripts/ui.ts (hunk 19 above)
Comment (m-cat) [reply scripts/ui.ts:489]: Yeah, we only want the promise to resolve when the right storage key is encountered. Any other storage key shouldn't trigger a `resolve`.

File: scripts/ui.ts (hunk 20 above)
Comment (m-cat) [reply scripts/ui.ts:493]: Added a comment.

File: scripts/ui.ts (hunk 19 above)
Comment (ro-tex) [reply scripts/ui.ts:491]: Should we `reject` here, then?

Comment (m-cat) [reply scripts/ui.ts:491]: No! 😅 We don't want login to fail if a different key is set for whatever reason. If there is another storage event listener for that key, it will also be triggered. Here, we are only interested in the specific key we are listening for.
